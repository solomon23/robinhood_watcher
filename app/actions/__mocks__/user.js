const tryAuthFunction = jest.fn()

export function tryAuth() {
  return tryAuthFunction
}

export default {
  tryAuth,
}
