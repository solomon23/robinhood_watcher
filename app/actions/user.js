import { authenticate, tryAuthenticate } from '../services/api'
import type { Dispatch } from '../reducers/types'

export const LOGIN_USER_LOADING = 'LOGIN_USER_LOADING'
export const LOGIN_USER_ERROR = 'LOGIN_USER_ERROR'
export const LOGIN_USER_SUCCESS = 'LOGIN_USER_SUCCESS'
export const LOGIN_USER_NEEDS_MFA = 'LOGIN_USER_NEEDS_MFA'
export const USER_LOGOUT = 'USER_LOGOUT'
export const SET_USER_ACCOUNT_NUMBER = 'SET_USER_ACCOUNT_NUMBER'

export function login(username, password, mfaCode) {
  return (dispatch: Dispatch) => {
    dispatch({ type: LOGIN_USER_LOADING })

    return authenticate(username, password, mfaCode)
      .then(response => {
        if (response.twoFactorAuthRequired && !mfaCode) {
          dispatch({ type: LOGIN_USER_NEEDS_MFA })
        } else if (response.authenticated) {
          dispatch({ type: LOGIN_USER_SUCCESS, response })
        } else {
          dispatch({ type: LOGIN_USER_ERROR, message: response.error })
        }

        return null
      }).catch(err => {
        dispatch({ type: LOGIN_USER_ERROR, error: err })
      })
    }
}

export function tryAuth() {
  return (dispatch: Dispatch) => {
    dispatch({ type: LOGIN_USER_LOADING })

    return tryAuthenticate()
      .then(response => {
        if (response.authenticated) {
          dispatch({ type: LOGIN_USER_SUCCESS, response })
        } else {
          dispatch({ type: LOGIN_USER_ERROR, message: response.error })
        }

        return null
      }).catch(err => {
        dispatch({ type: LOGIN_USER_ERROR, error: err })
      })
    }
}

export function logout() {
  return {
    type: USER_LOGOUT,
  }
}