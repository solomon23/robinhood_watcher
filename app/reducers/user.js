// @flow
import * as userActions from '../actions/user'
import * as appActions from '../actions/app'
import type { Action } from './types'

const defaultState = {
  isLoading: false,
  isError: false,
  errorReason: null,
  needsMfa: false,
  authenticated: false,
  accountNumber: null,
}

export default function app(state = defaultState, action: Action) {
  switch (action.type) {
    case userActions.LOGIN_USER_LOADING:
      return { ...defaultState, isLoading: true, authenticated: false }

    case userActions.LOGIN_USER_ERROR:
      return { ...defaultState, isError: true, errorReason: action.message }

    case userActions.LOGIN_USER_NEEDS_MFA:
      return { ...defaultState, needsMfa: true, authenticated: false }

    case userActions.LOGIN_USER_SUCCESS:
      return { ...defaultState, authenticated: action.response.authenticated }

    case appActions.GET_ACCOUNT_NUMBER.SUCCESS:
      return { ...state, accountNumber: action.response }

    case userActions.USER_LOGOUT:
      return { ...defaultState, authenticated: false }

    default:
      return state
  }
}
