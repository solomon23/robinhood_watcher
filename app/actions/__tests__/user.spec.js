import configureMockStore from 'redux-mock-store'
import fetchMock from 'fetch-mock'
import thunk from 'redux-thunk'
import * as actions from '../user'

const mockStore = configureMockStore([thunk])

describe('user actions', () => {
  afterEach(() => {
    fetchMock.restore()
  })

  it('should fire logout', () => {
    expect(actions.logout()).toMatchSnapshot()
  })

  it('should fire login', () => {
    fetchMock.mock('https://api.robinhood.com/oauth2/token/',
      { body: { access_token: 10, refresh_token: 20 } }
    )
    const store = mockStore({ user: { authenticated: false, accountNumber: null } })

    return store.dispatch(actions.login('joe', 'pass', 'mfa'))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
        return null
      })
  })

  it('should succeeded at trying to authenticate', () => {
    fetchMock.mock('https://api.robinhood.com/accounts/',
      { body: { results: {} } }
    )
    const store = mockStore({ user: { authenticated: false, accountNumber: null } })

    return store.dispatch(actions.tryAuth())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
        return null
      })
  })

  it('should fail at trying to authenticate', () => {
    fetchMock.mock('https://api.robinhood.com/accounts/', 401)
    fetchMock.mock('https://api.robinhood.com/oauth2/token/', 401)
    const store = mockStore({ user: { authenticated: false, accountNumber: null } })

    return store.dispatch(actions.tryAuth())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
        return null
      })
  })

  it('should succeeded at refreshing the token', () => {
    fetchMock.config.overwriteRoutes = false
    fetchMock.once('https://api.robinhood.com/accounts/', 401)
    fetchMock.once('https://api.robinhood.com/accounts/',
      { body: { results: {} } }
    )
    fetchMock.mock('https://api.robinhood.com/oauth2/token/',
      { body: { access_token: 10, refresh_token: 20 } }
    )
    const store = mockStore({ user: { authenticated: false, accountNumber: null } })

    return store.dispatch(actions.tryAuth())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot()
        return null
      })
  })
})
