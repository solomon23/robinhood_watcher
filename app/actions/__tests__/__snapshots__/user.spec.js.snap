// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user actions should fail at trying to authenticate 1`] = `
Array [
  Object {
    "type": "LOGIN_USER_LOADING",
  },
  Object {
    "message": undefined,
    "type": "LOGIN_USER_ERROR",
  },
]
`;

exports[`user actions should fire login 1`] = `
Array [
  Object {
    "type": "LOGIN_USER_LOADING",
  },
  Object {
    "response": Object {
      "authenticated": true,
      "twoFactorAuthRequired": false,
    },
    "type": "LOGIN_USER_SUCCESS",
  },
]
`;

exports[`user actions should fire logout 1`] = `
Object {
  "type": "USER_LOGOUT",
}
`;

exports[`user actions should succeeded at refreshing the token 1`] = `
Array [
  Object {
    "type": "LOGIN_USER_LOADING",
  },
  Object {
    "response": Object {
      "authenticated": true,
    },
    "type": "LOGIN_USER_SUCCESS",
  },
]
`;

exports[`user actions should succeeded at trying to authenticate 1`] = `
Array [
  Object {
    "type": "LOGIN_USER_LOADING",
  },
  Object {
    "response": Object {
      "authenticated": true,
    },
    "type": "LOGIN_USER_SUCCESS",
  },
]
`;
